sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpEventBatch.cs

```diff
@@ -157,6 +163,28 @@ public override bool TryAdd(EventData eventData)
                 _sizeBytes = size;
                 _batchMessages.Add(message);
 
+                // For a data body, the message converter will marshal the ReadOnlyMemory into an
+                // ArraySegment projection to avoid allocating.  To ensure the batch is immutable,
+                // it will need to own a buffer for the data body to guard against mutation of the
+                // source event.
+                //
+                // Because the batch is disposable and allows for deterministic clean-up, rent buffers
+                // from the shared pool rather than allocating a dedicated instance for each body segment.
+
+                if (eventData.GetRawAmqpMessage().Body.BodyType == AmqpMessageBodyType.Data)
+                {
+                   foreach (var bodySegment in message.DataBody)
+                   {
+                       var bodyArraySegment = (ArraySegment<byte>)bodySegment.Value;
+                       var buffer = ArrayPool<byte>.Shared.Rent(bodyArraySegment.Count);
+
+                       _rentedBuffers.Add(buffer);
+                       CopyArraySegmentTo(bodyArraySegment, buffer);
+
+                       bodySegment.Value = new ArraySegment<byte>(buffer, 0, bodyArraySegment.Count);
```

Comment:
Do we keep a reliable reference around for the collection of bodySegments anywhere? If so, I wonder if we could eliminate the need for the `_rentedBuffers` List since we can get a reference to the array reference that was used to construct the ArraySegment via its [Array](https://learn.microsoft.com/en-us/dotnet/api/system.arraysegment-1.array?view=net-6.0) property

Reply:
It's possible to do from the set of AMQP messages in the batch, but it's ugly with nested loops and casts:

```csharp
private static void ReturnRentedBuffers(List<AmqpMessage> batchMessages)
{
    foreach (var message in batchMessages)
    {
        if (((message.BodyType & SectionFlag.Data) > 0) && (message.DataBody != null))
        {
            foreach (var segment in message.DataBody)
            {
                try
                {
                    ArrayPool<byte>.Shared.Return(((ArraySegment<byte>)segment.Value).Array);
                }
                catch (ArgumentException)
                {
                    // Ignore.
                    //
                    // This occurs when the buffer passed is not rented. If this scenario
                    // is encountered, it likely indicates an error happened during TryAdd
                    // when renting the buffer or at Dispose.  The remainder of the rented
                    // buffers should still be returned.
                }
            }
        }
    }
}
```

The current implementation trades an allocation and a bit of overhead to amortize the clean-up cost.   Thoughts?

Reply:
I don't have much context as to whether or not the allocation savings are important enough in this case, but I was more curious if we had considered that as an option. I defer to your judgement on the tradeoff :)

Reply:
It's ugly, but after playing with it a bit, I decided to go that way and remove the need for the tracking list of rented buffers.